Feature: Option to group vehicle list by shared orders

btzy opened the pull request:
I added a "group by" dropdown to all vehicle lists to display one row per shared order, instead of one row per vehicle.  I've tested it with all vehicle types, as well as with the "vehicle groups" window.  When grouped by shared orders, clicking on a row will open the "shared orders of X vehicles" window.

Rationale:  Quite often, we are concerned only with "routes" instead of individual vehicles.  This feature highlights the important information (i.e. the routes) while reducing the importance of non-important information (i.e. the individual vehicles).

Extension:  It is possible to have a third option - to group by [vehicle group](https://wiki.openttd.org/Vehicle_groups).  As vehicle groups are mutually exclusive, it should be relatively simple to implement this extension.  However, I'm not sure of its usefulness and I didn't want to make this PR too big.

Note:  The new dropdown is available in the "shared orders of X vehicles" window too.  Though it is of not much use, I left it in for consistency.  The dropdown will be useful for the extension described above.

This is as per my post on TT Forums: https://www.tt-forums.net/viewtopic.php?f=32&t=84587&sid=fa68e80118896fca8f275a70d5cf6045

Some screenshots:
![image](https://user-images.githubusercontent.com/6948096/50831320-6950ec80-1385-11e9-818e-6aca1ee9cdc0.png)
![image](https://user-images.githubusercontent.com/6948096/50831357-8d143280-1385-11e9-922c-d8576936fd3b.png)
![image](https://user-images.githubusercontent.com/6948096/50831452-d5cbeb80-1385-11e9-9eaf-e10270b288b1.png)
![image](https://user-images.githubusercontent.com/6948096/50831551-2c392a00-1386-11e9-910b-be356dcb33b4.png)

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -1388,44 +1538,101 @@ void BaseVehicleListWindow::DrawVehicleListItems(VehicleID selected_vehicle, int
+					SetDParam(0, v->index);
+					DrawString(text_left, text_right, y, STR_TINY_BLACK_VEHICLE);
+				}
+				else if (v->group_id != DEFAULT_GROUP) {
```
codestyle: `} else if(..) {`

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -1388,44 +1538,101 @@ void BaseVehicleListWindow::DrawVehicleListItems(VehicleID selected_vehicle, int
+				if (v->IsChainInDepot()) {
+					str = STR_BLUE_COMMA;
+				}
+				else {
```
codestyle: `} else {`

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -1388,44 +1538,101 @@ void BaseVehicleListWindow::DrawVehicleListItems(VehicleID selected_vehicle, int
+				break;
+			}
+
+			case GB_SHARED_ORDERS: {
```
codestyle: no need for braces where no variable is declared

LordAro commented on src/widgets/group_widget.h:
```diff
@@ -15,6 +15,8 @@
 /** Widgets of the #VehicleGroupWindow class. */
 enum GroupListWidgets {
 	WID_GL_CAPTION,                  ///< Caption of the window.
+	WID_GL_GROUP_BY_ORDER,            ///< Group order.
```
Check alignment here, also remember to run `src/script/api/generate_widget.sh`

btzy replied:
Hmm this might be hard.  I develop primarily on Windows/MSVC.  Let me see how I can get this script to run...

EDIT: Done.  Git for Windows comes with workable bash.

LordAro commented on src/widgets/vehicle_widget.h:
```diff
@@ -63,7 +63,9 @@ enum VehicleDetailsWidgets {
 /** Widgets of the #VehicleListWindow class. */
 enum VehicleListWidgets {
 	WID_VL_CAPTION,                  ///< Caption of window.
+	WID_VL_GROUP_ORDER,               ///< Group order.
```
alignment

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -110,30 +166,70 @@ uint GetUnitNumberDigits(VehicleList &vehicles)
+			vehgroup->display_profit_last_year = std::accumulate(begin, end, static_cast<Money>(0), [](const Money &sum, const Vehicle* const &v) {
+				return sum + v->GetDisplayProfitLastYear();
+			});
+			vehgroup->age = std::accumulate(begin, end, static_cast<Date>(0), [](const Date &largest, const Vehicle* const &v) {
```
As much as I like the std::accumulate, perhaps a single loop would be nicer here, instead of 3?

LordAro reviewed (dismissed):
General idea is great, thanks for contributing! It's quite a large diff with lots of changes all together, do you think it would be feasible to split it into multiple logical commits?
I wonder about the fairly significant use of C++11 stuff as well. Technically we should be all fine for it, but...

btzy commented:
@LordAro Maybe I can split it into two commits:  The first commit would split the original `this->vehicles` into `this->vehicles` and `this->vehgroups`.  And the second commit would introduce `GB_NONE` and `GB_SHARED_ORDERS`.  Would this be okay?

btzy committed:
Codechange: Make vehicle lists internally support grouping of vehicles

This is in preparation for the new UI feature that allows grouping by shared orders.

btzy committed:
Feature: Option to group vehicle lists by shared orders

This applies to all kinds of vehicle lists, as well as the "vehicle groups" window.

btzy committed:
Cleanup: Move EndContainer() for empty WWT_PANEL to the same line

btzy committed:
Cleanup: Use range-based for-loop in group_gui.cpp

michicc commented on src/vehicle_gui.cpp:
```diff
@@ -110,30 +137,35 @@ uint GetUnitNumberDigits(VehicleList &vehicles)
 
-	this->unitnumber_digits = GetUnitNumberDigits(this->vehicles);
+	this->vehgroups.Clear();
+		uint max_unitnumber = 0;
```
Indentation.

michicc replied:
This is in relation to the first commit, it is correct after the second commit.

btzy replied:
Fixed.

michicc commented on src/vehicle_gui.cpp:
```diff
@@ -1388,17 +1420,22 @@ void BaseVehicleListWindow::DrawVehicleListItems(VehicleID selected_vehicle, int
+		DrawVehicleProfitButton(vehgroup.age, vehgroup.display_profit_last_year, vehgroup.NumVehicles(), vehicle_button_x, y + FONT_HEIGHT_NORMAL + 3);
+
+		const Vehicle *v = vehgroup.GetSingleVehicle();
+		StringID str;
```
If you are touching that anyway, please move the declaration down to just before the first use.

btzy replied:
Fixed.

michicc commented on src/vehicle_gui.cpp:
```diff
@@ -1420,12 +1457,40 @@ void BaseVehicleListWindow::DrawVehicleListItems(VehicleID selected_vehicle, int
 		DrawString(left, right, y + 2, str);
 
-		DrawVehicleProfitButton(v, vehicle_button_x, y + FONT_HEIGHT_NORMAL + 3);
+
```
Empty line.

michicc replied:
This is in relation to the first commit, it is correct after the second commit.

btzy replied:
Fixed.

PeterN commented:
Seems confusing to have both explicit groups and this. Not sure where they should fit together.

PeterN commented:
This doesn't compile currently:

> `src/vehicle_gui.cpp:204:73: error: initialized lambda captures are a C++14 extension
>        [-Werror,-Wc++14-extensions]
>                        const Vehicle **end = std::find_if_not(begin, this->vehicles.End(), [first_shared = (*...`

Plus it would be useful to rebase to master.

btzy commented:
> Seems confusing to have both explicit groups and this. Not sure where they should fit together.

My original motivation was that it was difficult to find a correct vehicle (for cloning / Ctrl-clicking) in a long list.  I tend to play with lots of vehicles, and without this patch I would have to click vehicles at random and open their order lists to see if I've gotten the correct one.

Grouping by shared orders also allows certain contextual actions (on the vehicle list) that explicit groups can't have, such as opening the "Shared orders of X vehicles" window, opening the order list, or Ctrl-clicking for cloning.

Since both explicit groups and shared orders already exist in OpenTTD and solve orthogonal concerns (explicit groups being arbitrarily user-defined and shared orders being defined by having the same route), maybe it would reduce confusion to have a third option in the "group by" dropdown that groups by the explicit (user-defined) group?  This would probably emphasise the distinction between explicit groups and shared orders.  (If this is done, maybe "user-defined group" is a better term than "vehicle group".)

Eddi-z commented:
I would maybe even go further and include other automatic group-like features like "all vehicles visiting \<station\>"

btzy commented:
> I would maybe even go further and include other automatic group-like features like "all vehicles visiting <station>"

I'm not sure how this could be a group-like feature.  Vehicles can visit multiple stations, so this kind of grouping doesn't seem to be mutually exclusive?  Looks more like a filter-like feature to me.

btzy commented:
> This doesn't compile currently:
> 
> > `src/vehicle_gui.cpp:204:73: error: initialized lambda captures are a C++14 extension [-Werror,-Wc++14-extensions] const Vehicle **end = std::find_if_not(begin, this->vehicles.End(), [first_shared = (*...`
> 
> Plus it would be useful to rebase to master.

Sorry about that, now fixed and rebased to master.

PeterN commented:
Thanks, now able to test it :-)

PeterN commented on src/group_gui.cpp:
```diff
@@ -632,26 +628,30 @@ class VehicleGroupWindow : public BaseVehicleListWindow {
 				this->SetDirty();
 				break;
 
+			case WID_GL_GROUP_BY_DROPDOWN:// Select grouping option dropdown menu
```
Missing space before comment.

PeterN commented on src/vehicle_gui_base.h:
```diff
@@ -35,19 +69,49 @@ struct BaseVehicleListWindow : public Window {
 	BaseVehicleListWindow(WindowDesc *desc, WindowNumber wno) : Window(desc), vli(VehicleListIdentifier::UnPack(wno))
 	{
-		this->vehicles.SetSortFuncs(this->vehicle_sorter_funcs);
+		// TODO: retrieve the user preference from somewhere instead of intializing to GB_NONE for every new window.
```
Could this be added? Probably need to ensure that grouped vehicle lists are still opened with GB_NONE.

PeterN commented:
Could the group-by stuff be hidden/disabled when in the grouped-by vehicle list?

btzy commented:
> Could the group-by stuff be hidden/disabled when in the grouped-by vehicle list?

I'm not sure which list you are referring to... do you mean the "Shared orders of X vehicles" list?

PeterN commented:
Yes, that particular one. I'm not sure if it makes sense or not to be have to have it grouped by shared orders, as it already is shared orders. On the other hand, if other group-by schemes are added it would again make sense. Hmm!

btzy commented:
Here's how it looks when disabled:

![image](https://user-images.githubusercontent.com/6948096/55018514-f83cdf00-502d-11e9-8515-8eb181aa6135.png)

btzy commented:
Now the grouping is saved (in memory, similar to the way the sort option is saved).  Different vehicle types have their grouping option saved independently, which is similar to the sort option.  Different vehicle list types also have their grouping option saved independently, because it seems reasonable that users might only want to group by shared orders when in particular lists (e.g. list of vehicles at a station).

Also, the grouping dropdown menu is disabled for the shared-orders vehicle list, because it probably isn't useful to be able to group a shared-orders vehicle list by shared orders.

btzy commented:
@PeterN I've made the changes as you suggested, and also changed a few things related to the recent commits that replaced `SmallVector` with ` std::vector`.

btzy commented:
@PeterN Can you please review this PR? Your stalebot is reminding me about this...

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -41,40 +41,49 @@
+
+/** Wrapper to convert a VehicleIndividualSortFunction to a VehicleGroupSortFunction */
+template <BaseVehicleListWindow::VehicleIndividualSortFunction func>
+static bool VehicleIndividualToGroupSorterWrapper(GUIVehicleGroup const &a, GUIVehicleGroup const &b) {
```
name seems backwards - this is a group to individual sorter, is it not?

btzy replied:
No, it converts (aka 'wraps') `func` (which is an individual sorter) to a version that will work with groups, so it's an individual to group sorter wrapper (to be read as "(individual to group) (sorter wrapper)").

But if you really want it written the other way round, I could change it.

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -41,40 +41,49 @@
+
+/** Wrapper to convert a VehicleIndividualSortFunction to a VehicleGroupSortFunction */
+template <BaseVehicleListWindow::VehicleIndividualSortFunction func>
+static bool VehicleIndividualToGroupSorterWrapper(GUIVehicleGroup const &a, GUIVehicleGroup const &b) {
```
also opening brace should be on its own line

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -41,40 +41,49 @@
-	&VehicleLengthSorter,
-	&VehicleTimeToLiveSorter,
-	&VehicleTimetableDelaySorter,
+BaseVehicleListWindow::VehicleGroupSortFunction * const BaseVehicleListWindow::vehicle_group_none_sorter_funcs[] = {
```
`vehicle_group_none_sorter_funcs` is a horrible name, we must be able to come up with something better

btzy replied:
It's for consistency with `vehicle_group_shared_orders_sorter_funcs`; these are modes of the 	`enum GroupBy`, which uses `GB_NONE` and `GB_SHARED_ORDERS`.  I can't think of a better name that would make sense here; could you provide a suggestion?

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -110,30 +144,37 @@ uint GetUnitNumberDigits(VehicleList &vehicles)
+		vehgroup.vehicles_end = it + 1;
+		vehgroup.display_profit_this_year = (*it)->GetDisplayProfitThisYear();
+		vehgroup.display_profit_last_year = (*it)->GetDisplayProfitLastYear();
+		vehgroup.age = (*it)->age;
```
seems like it would be nicer if GUIVehicleGroup had a constructor, so you didn't have to set all these values manually

btzy replied:
Agreed with this, I've added a constructor.

LordAro commented on src/vehicle_gui_base.h:
```diff
@@ -12,19 +12,53 @@
 
+struct GUIVehicleGroup {
+	VehicleList::const_iterator vehicles_begin;    ///< Pointer to beginning element of this vehicle group.
+	VehicleList::const_iterator vehicles_end;      ///< Pointer to past-the-end element of this vehicle group.
```
storing iterators like this scares me a little - there's no way to know if they've been invalidated

btzy replied:
They may be slightly dangerous, but I think it's justified in this scenario.

Only `BaseVehicleListWindow::BuildVehicleList()` will modify the underlying `VehicleList`, which will cause the iterators to be invalidated.  I'll add some comments around to make it clear that `vehgroups` requires stability of the underlying vehicle list.

I didn't want to make `GUIVehicleGroup` contain a `std::vector` because of these reasons:
- The current design reinforces the notion that the grouping option is just a "view" over the actual vehicle list.
- So that we can make the optimisation mentioned in `vehicle_gui.cpp:185` (`This can be optimized so that merely changing the "group by" dropdown value does not need to call GenerateVehicleSortList again.`) in the future, if found to be necessary.
- So that we won't have so many tiny vectors floating around the heap.

It might be possible to do something like:
```
template <typename T>
class GUIGroupableList<T> : public GUIList<T>;

typedef GUIVehicleGroupList GUIGroupableList<const Vehicle *>;
```
`GUIGroupableList` would contain the grouping mechanism, as well as the current `struct GUIVehicleGroup` (but perhaps made generic as appropriate).  This would effectively combine `BaseVehicleListWindow::vehicles` and `BaseVehicleListWindow::vehgroups`.  However, this would add quite a bit of code, and designing `GUIGroupableList` in a generic manner will require some kind of accumulation mechanism (e.g. to accumulate `display_profit_this_year`, `display_profit_last_year`, and `age`) (perhaps we have `GUIGroupList<U, Func, T>` where `U` is the accumulated type (i.e. a struct that contains `display_profit_this_year`, `display_profit_last_year`, and `age` fields), `T` is the original type (i.e. `const Vehicle *`), and `Func` is a binary function such that `U` is constructible from `std::invoke_result<Func, U, T>`).  This feels less easily understood as compared to the current design.

LordAro commented on src/vehicle_gui_base.h:
```diff
@@ -12,19 +12,53 @@
+	std::ptrdiff_t NumVehicles() const {
+		return std::distance(vehicles_begin, vehicles_end);
+	}
+	const Vehicle * GetSingleVehicle() const {
```
more opening braces in the wrong place (there's quite a few others, i won't bother pointing them out)

LordAro commented on src/vehicle_gui_base.h:
```diff
@@ -12,19 +12,53 @@
+	};
+
+	GroupBy grouping;                         ///< How we want to group the list.
+	std::vector<const Vehicle*> vehicles;     ///< List of vehicles.  This is the buffer for vehgroups to point into.
```
missing space before '*'
Looks like this line is replaced by VehicleList in the next commit, no reason why this couldn't be done here

LordAro commented on src/group_gui.cpp:
```diff
@@ -573,7 +573,7 @@ class VehicleGroupWindow : public BaseVehicleListWindow {
 				size_t vehicle_count = this->vehicles.size();
 
-				for (uint i = 0; i < vehicle_count; i++) {
+				for (uint32 i = 0; i < vehicle_count; i++) {
```
probably switch the whole thing out with a range-for loop, since you're changing it anyway

btzy commented:
@LordAro I'm a bit busy these few weeks, I'll take a look at your comments sometime soon!

LordAro commented:
Given the recent large codechange in #7864 , can you rebase and check everything still works? I'm almost surprised there isn't any conflicts...

EDIT: Well, there are definitely conflicts now

btzy commented:
@LordAro Fixed conflicts and rebased on master, this should be ready for re-review now.  I didn't implement a similar behaviour for Ctrl-clicking on a vehicle, because it is possible for two vehicles that share orders to be in different groups.

Note: I find it weird that in the group GUI, if I Ctrl-click a vehicle, the selected group changes, AND the vehicle view window is shown.  But in the station vehicle GUI, if I Ctrl-click a vehicle, the group GUI opens to the selected vehicle's group, but the vehicle view window doesn't open.  Seems like an inconsistency...  This isn't really related to this PR; I guess there might be reasons for this behaviour...

btzy commented:
Can someone look at this PR?  It has been sitting here more than a year now... @LordAro @PeterN

LordAro reviewed (dismissed):
You're right, this has been sitting around for far too long. I'd like to get an approval from a 2nd dev before merging though, just because of the size of the change.

TrueBrain commented:
Again, sorry for the long delay. I hope you are still interested in helping us get this merged.

I was about to merge it, but I spotted something my OCD couldn't handle. And I do apologize if it has been talked about before, but:

On some windows we have Sort -> Group, on others Group -> Sort. This feels really odd to me, especially if there is more than one of those windows open. Is there any reason for having them in different order depending on the window?

I know, it is rather minor. Other than that, this is good to merge as far as I am concerned :)

PS: I rebased your work so I could get the preview to work :) Hope you don't mind!

btzy commented:
@TrueBrain Thanks for reviewing.

This PR modifies the vehicle list windows by adding the "Group by" dropdown _above_ the existing "Sort by" dropdown.  This is the only way that makes sense, since the options for "Sort by" is dependent on the selected choice of "Group by".

Indeed, in the station windows, the "Sort by" dropdown is above the "Group by" dropdown, but I didn't want to change the existing behaviour since that window is unrelated to this PR.

btzy commented:
I looked at the station window again, and I think the "Sort by" there may arguably be a misnomer.
![image](https://user-images.githubusercontent.com/6948096/102310148-5b5e6a00-3fa5-11eb-819f-ac11648378c3.png)

The passengers shown in the window are displayed in a three-level nested list, and the nesting order is controlled by choosing one of the six permutations of the "Group by" dropdown.  The "Station: Waiting" vs "Amount: Waiting" options sort items in the same nesting level by station name (alphabetical order) or amount of passengers respectively.  However, "Station: Planned" shows a different number of passengers, presumably the amount of passengers who have decided to travel via this station but have not arrived yet.  Since the total amount of passengers changed, "Sort by" doesn't adequately capture the meaning of this dropdown.

I would argue to change the "Sort by" to something else for the station window, but perhaps that is outside the scope of this PR.

![image](https://user-images.githubusercontent.com/6948096/102310486-05d68d00-3fa6-11eb-8cd4-e5a6abaeac74.png)

![image](https://user-images.githubusercontent.com/6948096/102310466-fd7e5200-3fa5-11eb-9c02-7ae30d17e13a.png)

![image](https://user-images.githubusercontent.com/6948096/102310861-c8263400-3fa6-11eb-9873-8169c7eda9e8.png)

![image](https://user-images.githubusercontent.com/6948096/102310892-d8d6aa00-3fa6-11eb-969d-dd0c5873d75b.png)

It feels to me that since the "real" sort variants (i.e. "Station: Waiting" vs "Amount: Waiting") sort only sibling rows of the nested list, it is less important than the "Group by" options, which modify the structure of the nested list.  Maybe a three-level menu might capture hierarchy better:
Show: {Waiting, Planned}
Group by: {Source-Via-Destination, Source-Destination-Via, Via-Source-Destination, Via-Destination-Source, Destination-Source-Via, Destination-Via-Source}
Sort by: {Station name, Amount}

(but in any case, this should probably be for a different PR)

2TallTyler commented:
When vehicles are grouped by shared orders, it is impossible to click and drag vehicles into a group (such as moving them out of "ungrouped trains"). The shared orders quasi-group window opens immediately and moves focus to that window.

Is this intended behavior?

btzy commented:
@2TallTyler This behaviour is intended.  I did not implement that dragging in the group window (yet) because it seems to add quite a bit of code to make vehicle groups be selectable instead of individual vehicles.

The relevant code is here (if it doesn't work, it's group_gui.cpp lines 730-752 of the amended code):

https://github.com/OpenTTD/OpenTTD/pull/7028/files#diff-de05b496b75576b725425212e2b431ddddd38bdfcc724c2042ccf28a356744aeR730-R752

It may be worth trying to implement this (in a separate PR) though...

2TallTyler commented:
I don't quite understand that code, but in current trunk, control-click-dragging a vehicle moves its entire shared order group. Could you use that behavior? (perhaps selecting the first vehicle in the group, if you can't directly select all of them).

btzy commented:
> control-click-dragging a vehicle moves its entire shared order group

Wow, somehow I didn't notice that that worked!

btzy committed:
Feature: Drag-and-drop vehicles in group GUI for shared order groups

btzy commented:
I've implemented the suggestion by @2TallTyler, it was a small change.

TrueBrain commented:
> Indeed, in the station windows, the "Sort by" dropdown is above the "Group by" dropdown, but I didn't want to change the existing behaviour since that window is unrelated to this PR.

Oops, indeed, that `Group by` already exists. But it does leave us with an inconsistent UI (well .. more inconsistent then it was before your PR :P). I wonder if, for now, we should simply switch those two around to create a bit of consistency between the UIs .. that should be trivial to implement (I hope), and allows further work like the one you mention.

What are you thought on this? (feel free to disagree btw, I am just thinking out loud here).

btzy commented:
@TrueBrain

> What are you thought on this? (feel free to disagree btw, I am just thinking out loud here).

I agree that switching the "Group by" with "Sort by" dropdowns in the station window is strictly better than the current version, since the "correct" use of "Sort by" (i.e. by station name or by amount) should logically come after the grouping option.  Should I implement this in the same PR?

I'm still not fully convinced that my three-level menu suggestion is ideal though.  While being the most hierarchically accurate (in my opinion), it takes up another row of valuable screen space for something that most users will never use, or only use once... I, myself, when playing this game, have never changed the sorting option from "Station: Waiting", and have set the grouping option to "Via-Destination-Source" when cargodist was added and left it there ever since.

TrueBrain commented:
If you wouldn't mind adding the switching of the two buttons in this PR (as a separate commit) that would be lovely; avoids us forgetting about it or creating a new PR for only just that :)

We can debate in a later PR or Discussion what we should do with those dropdowns, as I agree that adding yet-another-button is not going to help out the UI :D

Tnx!

btzy committed:
Change: Place "Group by" above "Sort by" in station window for consistency

btzy commented:
@TrueBrain Done!

TrueBrain commented on src/vehicle_gui.cpp:
```diff
@@ -1611,27 +1810,47 @@ struct VehicleListWindow : public BaseVehicleListWindow {
+					case GB_SHARED_ORDERS:
+						assert(vehgroup.NumVehicles() > 0);
+						// TODO: support ctrl-click for cloning shared orders
+						// this might be hard, because there is currently no way to clone shared orders except by supplying a vehicle (but we don't want to enable other kinds of vehicle functionality)
```
I don't really like `TODO`s in code. It is an idea you currently have, which might faint over time. And it becomes more and more difficult over time to understand what people meant. So either we plan to add this, and do that, or we don't. But mentioning possible feature additions .. I suggest we just remove the comment ;)

TrueBrain replied:
You did not pick up this comment (and the similar other two); you still working on that, or you don't agree? :)

btzy replied:
Still working on that, didn't have the time to do it this morning :)

TrueBrain replied:
No worries, didn't want to rush you :) Just sometimes it is hard to see what is and isn't done :) Tnx!

btzy replied:
Done now :)

TrueBrain commented on src/vehicle_gui.cpp:
```diff
@@ -1468,6 +1660,10 @@ struct VehicleListWindow : public BaseVehicleListWindow {
 		if (this->vli.type == VL_SHARED_ORDERS) {
 			this->GetWidget<NWidgetCore>(WID_VL_CAPTION)->widget_data = STR_VEHICLE_LIST_SHARED_ORDERS_LIST_CAPTION;
+			// If we are in the shared orders window, then disable the group-by dropdown menu.
+			// Remove this when the group-by dropdown menu has another option apart from grouping by shared orders.
```
Nitpick: multi-line comments should use `/* */`

TrueBrain commented on src/vehicle_gui.cpp:
```diff
@@ -1611,27 +1810,47 @@ struct VehicleListWindow : public BaseVehicleListWindow {
 				this->SetDirty();
 				break;
 
+			case WID_VL_GROUP_BY_PULLDOWN:// Select sorting criteria dropdown menu
```
Nitpick: missing space between `:` and `//` (fix the `case` below too while at it please :D)

TrueBrain commented on src/vehicle_gui.cpp:
```diff
@@ -1384,44 +1535,99 @@ void BaseVehicleListWindow::DrawVehicleListItems(VehicleID selected_vehicle, int
+void BaseVehicleListWindow::UpdateVehicleGroupBy(GroupBy group_by)
+{
+	if (this->grouping != group_by) {
+		*this->sorting = this->vehgroups.GetListing(); // save the old sorting
```
This comment I do not understand. It states what it is doing, but not why. So it is confusing me, as now I am curious what it is designed to do :D Could you either make a comment that also explains the why a bit? Cheers!

btzy replied:
Added, hopefully it's clearer now.

TrueBrain replied:
Yes! That is exactly what I would like to read in a comment :) Tnx!

TrueBrain commented on src/vehicle_gui.cpp:
```diff
@@ -108,30 +171,63 @@ uint GetUnitNumberDigits(VehicleList &vehicles)
+		this->unitnumber_digits = CountDigitsForAllocatingSpace(max_unitnumber);
+	} else {
+		std::stable_sort(this->vehicles.begin(), this->vehicles.end(), [](const Vehicle * const &u, const Vehicle * const &v) {
+			return u->FirstShared() < v->FirstShared(); // Sort by the primary vehicle; we just want all vehicles that share the same orders to form a contiguous range.
```
Better put this comment on its own line in a `/* */`. In general, not a big fan of `//` behind lines comments .. only for small stuff it is fine.

TrueBrain commented on src/vehicle_gui.cpp:
```diff
@@ -108,30 +171,63 @@ uint GetUnitNumberDigits(VehicleList &vehicles)
 
+	this->vehgroups.clear();
+
+	/* Note: This can be optimized so that merely changing the "group by" dropdown value does not need to call GenerateVehicleSortList again. */
```
Not sure what I need to do with this comment ... should I expect you to pick this up? And I am pretty sure there are many many more places we can optimize things .. similar with some other comments, these comments only confuse us more in a year or in 2 years, than that they add value. I would suggest removing it.

btzy replied:
I guess... I put that there in case at some point we realise that GenerateVehicleSortList() runs too slowly... but will remove it.

TrueBrain commented on src/group_gui.cpp:
```diff
@@ -851,11 +875,32 @@ class VehicleGroupWindow : public BaseVehicleListWindow {
+						/*
+						No VehicleClicked(v) support for now, because don't want
+						to enable any contextual actions except perhaps clicking/ctrl-clicking to clone orders.
+						*/
```
Later in this patch this becomes a `TODO`. I explained a bit there why I find these comments odd .. "for now" , suggesting we can expect another PR soon? :)

(also another instance of this slightly above)

btzy replied:
Done

TrueBrain reviewed (changes requested):
Some minor things, nothing fancy. Almost there, I would say :)

Mostly, some comments confused me more than they helped me understand the code. Maybe you can look at those and see if you can add some letters (read: words) to make it more obvious what the intention was :)

Tnx!
